### ISSUE TITLE
Lowercase authorization header when using custom signature calculator

### ISSUE THREAD

**Issue by maxdietrich** [playframework/playframework#7735]
### Play Version (2.5.x / etc)

2.6.3

### API (Scala / Java / Neither / Both)

Scala

### Operating System (Ubuntu 15.10 / MacOS 10.10 / Windows 10)

MacOS 10.12.6

### JDK (Oracle 1.8.0_72, OpenJDK 1.8.x, Azul Zing)

1.8.0_131

### Expected Behavior

1. Send a request with a custom SignatureCalculator
2. Set the Authorization header in the signature calculator like this: `requestBuilder.addHeader(HeaderNames.AUTHORIZATION, "some value")`
3. Use the signature calculator like this in a unit test like described below.
4. One would expect the Authorization header to be written like in `HeaderNames.AUTHORIZATION` (i.e. "Authorization")

### Actual Behavior

When the test case below is executed, `authorization -> ...` is part of the printed Map, `Authorization` does not occur in it. The authorization header is written in lowercase.

### Reproducible Test Case

```
Server.withRouter() {
      case GET(p"/some-resource") => Action { request =>
        println(request.headers.toMap)
        Results.Ok("")
      }
    } { implicit port =>
      WsTestClient.withClient { client =>
        val resultFuture = client
          .url(s"http://localhost:$port/some-resource")
          .sign(mySignatureCalculator)
          .get()

        whenReady(resultFuture) { _.status mustBe OK }
      }
    }
```

**Comment by richdougherty** [comment]
We'll need to investigate this to find out whether the header name is changed by Play, by Akka HTTP or by AsyncHttpClient. We should be try to preserve header case if possible.

**Comment by richdougherty** [comment]
Suggestion for a fix/workaround: https://github.com/playframework/playframework/issues/7737#issuecomment-323204391

**Comment by dietmax** [comment]
Hey, are there any news on this? We would really like to migrate to Play 2.6 and this blocks us from doing so right now.

**Comment by gmethvin** [comment]
@dietmax Perhaps I don't understand the problem fully, but if the problem is Play's akka-http server as we suspect it should not be a blocker in updating. You should be able to switch to the Netty server (which is still 100% supported), or you can just use the standard APIs for getting headers, which are case insensitive (e.g. `request.headers.get(AUTHORIZATION)`).

I agree it's a good idea to preserve the case here though.

**Comment by schmitch** [comment]
> I agree it's a good idea to preserve the case here though.

well basically it might be a good idea, but actually http headers are case-insensitive. This matters a lot on http2 where a lot of implementations do lowercase every header even on http/1.1 for forward compatibility.

**Comment by dietmax** [comment]
@gmethvin Switching to the netty server is definitely something we will look into, thanks for that tip 👍 Regarding the standard API for getting headers, my understanding is, that `request.headers` is just a wrapper around a `Map[String, String]` and HeaderNames.AUTHORIZATION is the String "Authorization". Therefore `request.headers.get("Authorization")` will only return Uppercase authorization headers, which is the behavior that causes my problem. Or am I missing something here?

**Comment by gmethvin** [comment]
@dietmax The `Headers` class internally uses a `Map` but it is a `TreeMap` with a [case insensitive ordering](https://github.com/playframework/playframework/blob/2.6.x/framework/src/play/src/main/scala/play/api/mvc/Headers.scala#L101). Headers are case-insensitive, so the headers API is case insensitive.

**Comment by dietmax** [comment]
Thanks to @gmethvin s comment I finally came around to the fact that `request.headers.keys.contains(HeaderNames.AUTHORIZATION)` does not behave in the same way that `request.headers.get(HeaderNames.AUTHORIZATION).isDefined` does. Transforming the `TreeMap` to a `Set` works around the logic that enables case insensitivity. That is where our error came from. Thank you!

**Comment by rtyley** [comment]
We've also observed that when using the default Akka-Http engine in Play 2.6, the header name will be lower-cased if it's 'invalid' in some way. In particular, for two HTTP requests supplying `User-Agent` (ie capitalised), the headers map will have either `User-Agent` or `user-agent` depending on whether Akka-HTTP things the User-Agent string is valid:

With a valid `User-Agent`:
```
curl -H "User-Agent: Mozilla/5.0 (iPhone; CPU iPhone OS 11_2_2 like Mac OS X) AppleWebKit/604.4.7 (KHTML, like Gecko) Mobile/15C202" localhost:9000/test

headers.toSimpleMap=Map(Host -> localhost:8970, Accept -> */*, User-Agent -> Mozilla/5.0 (iPhone; CPU iPhone OS 11_2_2 like Mac OS X) AppleWebKit/604.4.7 (KHTML, like Gecko) Mobile/15C202, Content-Type -> none/none, Remote-Address -> 127.0.0.1:44698, ...)
```


With an invalid `User-Agent`:
```
$ curl -H "User-Agent: Mozilla/5.0 (iPhone; CPU iPhone OS 11_2_2 like Mac OS X) AppleWebKit/604.4.7 (KHTML, like Gecko) Mobile/15C202 [FBAN/FBIOS;FBAV/155.0.0.36.93;FBBV/87992437;FBDV/iPhone9,3;FBMD/iPhone;FBSN/iOS;FBSV/11.2.2;FBSS/2;FBCR/3Ireland;FBID/phone;FBLC/en_US;FBOP/5;FBRV/0]" localhost:9000/test

[WARN] [01/24/2018 15:21:13.122] [play-dev-mode-akka.actor.default-dispatcher-10] [akka.actor.ActorSystemImpl(play-dev-mode)] Illegal header: Illegal 'user-agent' header: Invalid input '[', expected 'EOI', product-or-comment, WSP, comment or CRLF (line 1, column 112): Mozilla/5.0 (iPhone; CPU iPhone OS 11_2_2 like Mac OS X) AppleWebKit/604.4.7 (KHTML, like Gecko) Mobile/15C202 [FBAN/FBIOS;FBAV/155.0.0.36.93;FBBV/87992437;FBDV/iPhone9,3;FBMD/iPhone;FBSN/iOS;FBSV/11.2.2;FBSS/2;FBCR/3Ireland;FBID/phone;FBLC/en_US;FBOP/5;FBRV/0]

headers.toSimpleMap=Map(Host -> localhost:8970, Accept -> */*, user-agent -> Mozilla/5.0 (iPhone; CPU iPhone OS 11_2_2 like Mac OS X) AppleWebKit/604.4.7 (KHTML, like Gecko) Mobile/15C202 [FBAN/FBIOS;FBAV/155.0.0.36.93;FBBV/87992437;FBDV/iPhone9,3;FBMD/iPhone;FBSN/iOS;FBSV/11.2.2;FBSS/2;FBCR/3Ireland;FBID/phone;FBLC/en_US;FBOP/5;FBRV/0], Content-Type -> none/none, Remote-Address -> 127.0.0.1:44116, ...)
```

Lower-case for invalid, capitalised for valid! Where is that documented?!

**Comment by richdougherty** [comment]
I'm working on Akka HTTP backend header issues right now. I'm going to re-open this issue so we can track it.

**Comment by richdougherty** [comment]
@maxdietrich, I haven't been able to replicate this issue. I.e. whenever I send `Authorization: some value` it stays as `Authorization: some value`. Do you have a curl command I could use to replicate?

Note: I do see the other way around, where `authorization: some value` becomes `Authorization: some value`. I will fix this.

@rtyley I have replicated your example on master and a fix is coming.

**Comment by richdougherty** [comment]
PR has been tidied up and more issues have been verified as being fixed. Now waiting on the release of Akka HTTP which should happen in the next few days.

**Comment by marcospereira** [comment]
@maxdietrich and @rtyley,

There is a fix for this issue on [Play 2.6.13](https://discuss.lightbend.com/t/play-2-6-13-released/629). Could you please try this new version and confirm it is working for you guys?

I'm closing it now, but if the fix doesn't work, we can reopen this issue with the additional information.

Thanks!
